libs/constrained-generators/constrained-generators.cabal

```diff
@@ -0,0 +1,65 @@
+cabal-version:      3.0
+name:               constrained-generators
+version:            0.2.0.0
+license:            Apache-2.0
+maintainer:         operations@iohk.io
+author:             IOHK
+synopsis:
+    Framework for generating constrained random data using
+    a subset of first order logic
+
+build-type:         Simple
+extra-source-files: CHANGELOG.md
+
+source-repository head
+    type:     git
+    location: https://github.com/input-output-hk/cardano-ledger
+    subdir:   libs/constrained-generators
+
+library
+    exposed-modules:
+        Constrained.Base
+        Constrained.Core
+        Constrained.Env
+        Constrained.Instances
+        Constrained.Spec
+        Constrained.List
+        Constrained.GenT
+        Constrained.Univ
+        Constrained.Graph
+        Constrained.Spec.Generics
+        Constrained.Spec.Pairs
+        Constrained.Spec.Maps
+        Constrained.Test
```

Comment:
I think it'd be nice to have some simple API for using the constrained generators, right now everything is exported including a lot of internal functions which require a deep understanding of the generators. 

As a user I'd be interested in writing predicates, calculating a spec from the predicates and then using the spec to either generate values or test some existing value against the predicates. Then the internal stuff could be added as hidden modules. Do you think it'd be viable to have such an API?

Reply:
There is now a module called `Constrained` that aims to be the _minimal_ interface you need to work comfortably with this stuff. However, that still exposes a lot of details that could be better handled with a slightly nicer interface.